wordpress-mobile/WordPress-iOS
# Share Extension for WordPress.com sites

#### Description:

Implements an iOS Share Extension using the built-in UI from iOS to provide a basic experience supporting text and URLs for WordPress.com-hosted and Jetpack sites. We don't support self-hosted non-Jetpack sites at the moment because the API calls are extremely complex and the share extension is a separate application with very little connection to the main app.

Closes #2231

#### Sample Screenshot:
![2016-03-08_13-13-24](https://cloud.githubusercontent.com/assets/373903/13612950/957bc2b2-e52f-11e5-9824-e6acf7b6331b.png)

#### First Iteration Features:
1. Share target for apps that provide text and URLs.
2. Post status can be publish or draft.
3. Site picker shows list of sites within WordPress.com world (including Jetpack).
4. Text is split between subject and body of post if multi-line.
5. URLs are automatically appended to the end of the post upon submission.

Future features will include support for media (images, video), better caching of site lists, remembering last used post status and site.

Like the today widget the provisioning profile setup is a bit stabby. We have another app ID for the share extension and another set of provisioning profiles to support it. (Sorry @sendhil :smile:)

#### To test:
1. Delete the app from your device.
2. Install fresh.
3. Attempt to use the share extension from Safari or News.app. You should receive a notice saying no site configured.
4. Log into the app with a self-hosted site only (no Jetpack connection).
5. Attempt to use the share extension. You should again receive a message indicating no site configured (because of .com support only).
6. Connect to WordPress.com in the app.
7. Attempt to use the share extension. You should be able to post to the site selected.

#### Additionally test:
* Site picker - ensure only visible .com sites are showing up. Your default site should be first in the list.
* Post status - ensure the post is posted with the right status.
* Maybe try to break all the things with emoji?

Needs review: @SergioEstevao, @aerych, @koke, @mattmiklic 


--


#### Pendings:

- [x] Signing out of the app, then signing back in can temporarily glitch things. [Details Here](https://github.com/wordpress-mobile/WordPress-iOS/pull/4928#issuecomment-194004525)
- [x] When you tap on post status, it seems you can't tap on Publish. Well, sometimes you can. [Details Here](https://cloudup.com/cTkj7wy0WII)
- [x] Nuke the "Legacy Share" Functionality.
- [x] Discuss and (perhaps / maybe / definitely) rename to WordPressShareExtension.
- [x] Verify if the Extension breaks in the odd case where you can configure the Today widget for a Jetpack site connected to a second wp.com account.
- [x] Verify the "Empty Content" Scenario.


#### SpinOff Issues:

- #4955: Signing into Jetpack (through stats) doesn't store the shared credentials (probably broken for the today widget too). [Details Here](https://github.com/wordpress-mobile/WordPress-iOS/pull/4928#issuecomment-193962073).
- #4957: Add Metrics to both, Share Extension and Today Widget.
